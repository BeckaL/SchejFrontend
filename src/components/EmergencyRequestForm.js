import React, { Component } from 'react'
import axios from 'axios'

export default class EmergencyRequestForm extends Component {
  constructor(props){
    super(props)
    this.state = {
      comment: ''
    }
  }
  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit = (event) => {
    event.preventDefault()
    axios.post(
      `${process.env.REACT_APP_API_URL}/emergency_requests`, {
        emergency_shift_id: this.props.shiftInfo.shiftId,
        comment: this.state.comment
      })
      .then(response => {
        this.props.history.push('/')
        this.props.history.push('/shifts')
      })
    }

  render() {
    return(
      <form id="emergency-request-form" onSubmit={this.handleSubmit}>
        <label className='shift-info-label' for='comment'>Comment:</label>
        <input className='shift-info-entry' id="comment" type="text" name="comment" placeholder="Why do you need to swap?" onChange={this.handleChange} /><br /><br />
        <section id='confirm-emergency'>
          <button id="confirm-emergency-request" className='custom-button'>Confirm</button>
        </section>
      </form>
    )
  }
}
