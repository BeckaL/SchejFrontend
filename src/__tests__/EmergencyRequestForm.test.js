import React from 'react'
import { create } from "react-test-renderer"
import { mount, shallow } from 'enzyme'
import axios from 'axios'
import EmergencyRequestForm from '../components/EmergencyRequestForm'


jest.mock('axios')

describe('Emergency Request Form', () => {
  var wrapper

  beforeAll(() => {
    wrapper = shallow(<EmergencyRequestForm history = {[]} shiftInfo = {{
      shiftId: 1
    }}/>)
  })

  test("it matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot()
  })

  test("it renders a form", () => {
    const form = wrapper.find("form")
    expect(form.length).toEqual(1)
  })

  test("it renders a button on the form", () => {
    const button = wrapper.find("button")
    expect(button.length).toEqual(1)
  })

  test("it makes correct api request when the button is clicked", () => {
    axios.post.mockResolvedValue({})
    const form = wrapper.find("form")
    const fakeEvent = { preventDefault () {} }
    form.props().onSubmit(fakeEvent)
    expect(axios.post).toHaveBeenCalledTimes(1)
  })

  test("it sends the correct parameters in the api request", () => {
    axios.post.mockResolvedValue({})
    const form = wrapper.find("form")
    const comment = wrapper.find("input")
    comment.simulate('change', { target: { value: 'this is a comment' } })
    const fakeEvent = { preventDefault () {} }
    form.props().onSubmit(fakeEvent)
    setTimeout(() => {
      expect(axios.post).toHaveBeenCalledWith(
        'http://localhost:3001/api/v1/emergency_requests', {
          comment: "this is a comment",
          emergency_shift_id: 1
        }
      )
    }, 20)
  })
})
