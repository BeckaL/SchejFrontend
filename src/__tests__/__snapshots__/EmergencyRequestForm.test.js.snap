// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emergency Request Form it matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EmergencyRequestForm
    history={Array []}
    shiftInfo={
      Object {
        "shiftId": 1,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          className="shift-info-label"
          for="comment"
        >
          Comment:
        </label>,
        <input
          className="shift-info-entry"
          id="comment"
          name="comment"
          onChange={[Function]}
          placeholder="Why do you need to swap?"
          type="text"
        />,
        <br />,
        <br />,
        <section
          id="confirm-emergency"
        >
          <button
            className="custom-button"
            id="confirm-emergency-request"
          >
            Confirm
          </button>
        </section>,
      ],
      "id": "emergency-request-form",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Comment:",
          "className": "shift-info-label",
          "for": "comment",
        },
        "ref": null,
        "rendered": "Comment:",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "shift-info-entry",
          "id": "comment",
          "name": "comment",
          "onChange": [Function],
          "placeholder": "Why do you need to swap?",
          "type": "text",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="custom-button"
            id="confirm-emergency-request"
          >
            Confirm
          </button>,
          "id": "confirm-emergency",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Confirm",
            "className": "custom-button",
            "id": "confirm-emergency-request",
          },
          "ref": null,
          "rendered": "Confirm",
          "type": "button",
        },
        "type": "section",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            className="shift-info-label"
            for="comment"
          >
            Comment:
          </label>,
          <input
            className="shift-info-entry"
            id="comment"
            name="comment"
            onChange={[Function]}
            placeholder="Why do you need to swap?"
            type="text"
          />,
          <br />,
          <br />,
          <section
            id="confirm-emergency"
          >
            <button
              className="custom-button"
              id="confirm-emergency-request"
            >
              Confirm
            </button>
          </section>,
        ],
        "id": "emergency-request-form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Comment:",
            "className": "shift-info-label",
            "for": "comment",
          },
          "ref": null,
          "rendered": "Comment:",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "shift-info-entry",
            "id": "comment",
            "name": "comment",
            "onChange": [Function],
            "placeholder": "Why do you need to swap?",
            "type": "text",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="custom-button"
              id="confirm-emergency-request"
            >
              Confirm
            </button>,
            "id": "confirm-emergency",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Confirm",
              "className": "custom-button",
              "id": "confirm-emergency-request",
            },
            "ref": null,
            "rendered": "Confirm",
            "type": "button",
          },
          "type": "section",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
